fonts/generate.py

```diff
@@ -0,0 +1,657 @@
+        return False
+
+    if not os.access(header_out_path, os.W_OK):
+        log.error(f"Could not write to {header_out_dir}.")
+        return False
+
+    log.debug("SMuFL header will be stored in %s", header_out_path)
+    supported_glyphs: dict = __get_supported_glyph_codes(supported_glyphs_fname)
+
+    fmt_supported_glyphs: list = [
+        f"    SMUFL_{gcode}_{gname} = 0x{gcode}," for gcode, gname in supported_glyphs.items()
+    ]
+
+    log.debug("SMuFL header will contain %s supported glyphs", len(fmt_supported_glyphs))
+    fmt_glyph_list = "\n".join(fmt_supported_glyphs)
+    fmt_header = SMUFL_HEADER.format(
+        smufl_glyph_list=fmt_glyph_list, len_smufl_codes=len(fmt_supported_glyphs)
+    )
+
+    log.debug("Writing %s", header_file_path)
+    with open(header_file_path, "w") as header_inc:
+        header_inc.write(fmt_header)
+
+    log.debug("Finished writing SMuFL header")
+    return True
+
+
+def extract_fonts(opts: Namespace) -> bool:
+    """
+    Takes a font file and extracts the necessary files for Verovio. Generates a glyph file for each glyph, and an SVG
+    glyph file containing only the supported SMuFL glyphs.
+
+    :param opts: A set of options from the argument parser sub-command.
+    :return: True if successful, False otherwise.
+    """
+    source_dir: str = opts.source
+    fontname: str = opts.fontname
+    data_dir: str = opts.data
+
+    source_pth: Path = Path(source_dir)
+    font_data_pth: Path = Path(source_pth, fontname)
+    metadata_pth: Path = Path(font_data_pth, f"{fontname.lower()}_metadata.json")
+    font_pth: Path = Path(source_pth, font_data_pth, f"{fontname}.svg")
+    data_pth: Path = Path(data_dir)
+    glyph_file_pth: Path = Path(data_pth, fontname)
+    output_pth: Path = Path(data_pth, f"{fontname}.xml")
+
+    log.debug("Extracting fonts for %s from %s", fontname, font_data_pth.resolve())
+
+    if not font_data_pth.is_dir() or not os.access(font_data_pth, os.R_OK):
+        log.error(f"Could not read font information from {str(font_data_pth)}")
+        return False
+
+    if not os.access(metadata_pth, os.R_OK):
+        log.error(f"Could not read {str(metadata_pth)}. Does it exist?")
+        return False
+
+    if not os.access(font_pth, os.R_OK):
+        log.error(f"Could not read {str(font_pth)}. Does it exist?")
+        return False
+
+    if not os.access(output_pth, os.W_OK):
+        log.error(f"Could not write to {str(output_pth)}. Check permissions.")
+        return False
+
+    if not glyph_file_pth.is_dir():
+        log.debug("Creating %s", glyph_file_pth)
+        os.makedirs(glyph_file_pth, exist_ok=True)
+
+    svg_data = __read_svg_font_file(str(font_pth))
+    if not svg_data:
+        log.error("Could not extract SVG data from %s", font_pth.resolve())
+        return False
+
+    family, units, hax, glyphs = svg_data
+
+    log.debug(
+        "SVG Data extracted: Family: %s, Units: %s, H-A-X: %s, Number of glyphs: %s",
```

Comment:
Why avoid an f-string here?

Reply:
In the `log` module, strings using the `%s` statements are lazily evaluated, while `f""` statements are not. So if your log level is `info` then percent-formatted strings do not actually get interpolated in `debug` statements.

See: https://medium.com/swlh/why-it-matters-how-you-log-in-python-1a1085851205

Reply:
Pylint will flag this, which is how I learned about it.